CS3219-AY2223S1/cs3219-project-ay2223s1-g12
# Add change password UI

Fixes #64 

Summary:
- Users might want to change their passwords
- Create a button component that redirects to change password page
- Change password page contains 3 text field inputs - username, old password, new password
- custom hook - `setSessionStorage` is created to make use of session storage to persist username information
- `username` field is taken from session storage, and set to disabled
- Linked to backend `changePassword` API using axios
- Display error using `error` prop in `textfield` and `snackbar`
- Add client-side validation for password strength

Issue 1:
- After backend login, `navigate` to `/home` without waiting for synchronous saving to session storage
- Workaround: make use of existing `state` in `useLocation` to store `username` and save to session storage when reached `home` page
- Causes an issue if manually redirect to `/home` page since there is no `state` variable present

![image](https://user-images.githubusercontent.com/65227564/193727545-4372d8b5-7537-48cb-a91d-03ec1dd374ec.png)
![image](https://user-images.githubusercontent.com/65227564/193727577-daf0fa4e-8694-44f5-a3a4-6d5f02fa70b6.png)
From the above images:
Notice that `username from login` does not appear in console at all
![image](https://user-images.githubusercontent.com/65227564/193727476-66b98d74-7eee-4fbf-833d-5e7cc08db7d3.png)

Issue 2: Resolved
- Bug in `authenticateCookieToken` in backend
- pending #78 
- workaround: removed `authenticateCookieToken` when `changePassword` API is called

How to test:
1. `npm i` necessary dependencies
2. `npm run dev` the user-service and `npm start` frontend
3. Go to `/login` page frontend and login. You should see that your `username` is displayed on the home page after login is successful.
4. Click `Change Password` Button. You should see this screen
![image](https://user-images.githubusercontent.com/65227564/193729278-839d137a-c23b-450f-9ab1-f3c8490da257.png)
5. Attempt to change your password
6. Errors will appear either as a red outlined text field or snackbar. Possible errors covered:
-- old password same as new password
-- new password does not meet password strength requirement
-- double confirm password does not match new password 
10. Click `Submit` to change password. If successful, a snackbar with success message appears.
![image](https://user-images.githubusercontent.com/65227564/193729052-5a7434ec-b1a0-4805-bd8b-0901efd9af00.png)
